src/io.rs

```diff
@@ -0,0 +1,59 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use crate::result::Result;
+
+#[derive(Debug)]
+pub(crate) struct Sd {
+    pub(crate) ptr: *const u8,
+    pub(crate) len: usize,
+}
+
+impl Sd {
+    pub(crate) fn new(ptr: *const u8, len: usize) -> Sd {
+        Sd { ptr, len }
+    }
+
+    pub(crate) fn from_slice(bs: &[u8]) -> Sd {
+        Sd::new(bs.as_ptr(), bs.len())
+    }
+
+    pub(crate) fn as_ptr(&self) -> *const u8 {
+        self.ptr
+    }
+
+    pub(crate) fn len(&self) -> usize {
+        self.len
+    }
+
+    pub(crate) fn subset(&self, offset: usize, len: usize) -> Sd {
+        assert!(offset + len <= self.len);
+        let ptr = self.ptr.wrapping_add(offset);
+        Sd { ptr, len }
+    }
+}
+
+pub(crate) trait Read {
+    fn read(&self, off: u64, dst: &mut [u8]) -> Result<usize>;
+    fn size(&self) -> usize;
+}
+
+impl Read for &[u8] {
+    fn read(&self, off: u64, dst: &mut [u8]) -> Result<usize> {
+        let off = off as usize;
+        if off > self.len() {
```

Comment:
... i did not know that `&slice[slice.len()..]` would work, but apparently it does and produces an empty slice!

Reply:
Indeed! It occurs to me that, regardless of correctness, it's clearer all around to just use `>=` here.